specgen/specgen/hellocap/listener.go

```diff
@@ -0,0 +1,200 @@
+package hellocap
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+)
+
+const tlsRecordHeaderLen = 5
+
+type contextKey int
+
+// contextHelloKey is the context key for retrieving the ClientHello.
+const contextHelloKey contextKey = 0
+
+// FromRequest returns the TLS ClientHello sent by the remote. This request must have come through a
+// handler attached to the Server type in this package.
+func FromRequest(req *http.Request) ([]byte, error) {
+	hello, ok := req.Context().Value(contextHelloKey).(capturedHello)
+	if !ok {
+		return nil, errors.New("no attached ClientHello")
+	}
+	return hello.hello, hello.err
+}
+
+// attachHello is used to put a captured ClientHello on a http.Request context. This is how the
+// ClientHello, captured at the TCP level, is made available to users of this library.
+func attachHello(req *http.Request, hello capturedHello) *http.Request {
+	ctx := context.WithValue(req.Context(), contextHelloKey, hello)
+	return req.Clone(ctx)
+}
+
+// onHello is a callback invoked when a ClientHello is captured. Only one of hello or err will be
+// non-nil. This callback is not invoked for incomplete ClientHellos. In other words, err is non-nil
+// only if what has been read off the connection could not possibly constitute a ClientHello,
+// regardless of further data from the connection.
+type onHello func(hello []byte, err error)
+
+// capturingConn is a TCP connection and implements net.Conn. If a TLS ClientHello is sent by the
+// peer, capturingConn will invoke the provided onHello callback. This is the basic building block
+// by which we build a hello-capturing server.
+type capturingConn struct {
+	// Wraps a TCP connection.
+	net.Conn
+
+	helloRead bool
+	helloBuf  *bytes.Buffer
+	helloLock sync.Mutex // protects fields in this block
```

Comment:
I'm not sure I see what this lock gives us. Let's assume for a second we wanted to run this in production, and generate lots of hellos for analysis: locking on `Read` would be pretty expensive. I think you maybe just don't need it at all, though? It seems unlikely that `net/http` would be calling `Read` concurrently on the same connection.

Reply:
You're correct that (a) locking is expensive and (b) concurrent reads are unlikely.  However, this is technically necessary as the `net.Conn` interface allows for concurrent reads.  Removing the lock would, for example, cause this implementation to fail `nettest.TestConn`.

Reply:
Hm, can we read the hello right after `Accept`, before we pass the connection anywhere? Then we haven't passed the conn anywhere yet, so it should be safe to skip the locking. Actually, come to think of it, we don't even need to override `Read`, which would void a whole bunch of code. We would need to pass an `io.MultiReader` back into the stdlib, though, with whatever we've already read prepended. And that might require a custom type.

Feel free to ignore all this if I'm overprescribing what is basically lab code. I find the exercise of tightening up i/o code to be enjoyable, but I know not everyone does.